### PR TITLE
feat: add time template helpers

### PR DESCRIPTION
**Pull request by freak12techno**
Copied from here https://github.com/prometheus/prometheus/blob/34ee8c607809cca973a1d7383713035045f681d8/template/template.go#L267, the eventual goal is to reuse it in Alertmanager as well to allow it as a template helper, and also in Prometheus.

Related: https://github.com/prometheus/alertmanager/issues/3717
Related: https://github.com/prometheus/alertmanager/pull/3720#issuecomment-1968901170

Function itself is moved as is, tests are updated as there's no templates rendering here.

Also ~~added the `.idea` to `.gitignore` (as I use Goland IDE and it creates some files that are probably not to be committed) and~~ added the go.mod entry to go.mod and go.sum (otherwise `make test` would fail).

### CODE REVIEW

**Comment by freak12techno** [comment]
@gotjosh can you look at this?

File: .gitignore [hunk 1]
```diff
@@ -1 +1,2 @@
-vendor/
\ No newline at end of file
+vendor/
+.idea
```
Comment (beorn7) [inline .gitignore:2]: Please don't add IDE specific entries to .gitignore. You have your own global gitignore files for that.

Comment (freak12techno) [reply .gitignore:2]: Okay, will remove, wait a sec.

Comment (freak12techno) [reply .gitignore]: Fixed, will also check what's up with go.mod

Comment (freak12techno) [reply .gitignore]: go.mod thing also seems to be fixed, although I do not quite get why this keeps happening.

**Comment by LMantovan** [comment]
Hi, in the old PR https://github.com/prometheus/alertmanager/pull/3720 you also included time.since. Is it possible to add here?

**Comment by freak12techno** [comment]
@LMantovan I'll still need to make a PR towards alertmanager to use this template, my plan is to add time.Since in the next PR which will supersede the one I closed.

**Comment by freak12techno** [comment]
@gotjosh ping?

**Comment by beorn7** [comment]
@gotjosh are you up to reviewing this? Either way, could you comment so that we at least know you have seen this PR?

**Comment by gotjosh** [comment]
I apologise for missing the notifications for these - like everyone else, managing Github notifications has become an impossible task at this point.

@grobinson-grafana do you have any spare cycles to review this?

File: helpers/templates/convert_to_float.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// Copyright 2015 The Prometheus Authors
```
Comment (gotjosh) [inline helpers/templates/convert_to_float.go:1]:
```suggestion
// Copyright 2024 The Prometheus Authors
```

File: helpers/templates/convert_to_float.go
Comment (gotjosh) [inline helpers/templates/convert_to_float.go:1]: I think we should have a single file containing everything, we call it `time.go` under `helpers/template/time.go` then we can have the tests file adjusted to be `time_test.go`.

File: helpers/templates/humamize_duration_test.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+func TestHumanizeDurationSecondsFloat64(t *testing.T) {
+	input := []float64{0, 1, 60, 3600, 86400, 86400 + 3600, -(86400*2 + 3600*3 + 60*4 + 5), 899.99}
+	expected := []string{"0s", "1s", "1m 0s", "1h 0m 0s", "1d 0h 0m 0s", "1d 1h 0m 0s", "-2d 3h 4m 5s", "14m 59s"}
+
+	for index, value := range input {
+		result, err := HumanizeDuration(value)
+		require.NoError(t, err)
+		require.Equal(t, expected[index], result)
+	}
```
Comment (gotjosh) [inline helpers/templates/humamize_duration_test.go:30]: nit: The test themselves are lovely, great cases.

It's more idiomatic in go (and even better for IDEs) if we structure the tests as a table. As such, I think a better way to structure your tests is:

```golang
	tc := []struct {
		name     string
		input    float64
		expected string
	}{
		{name: "zero", input: 0, expected: "0s"},
		{name: "one second", input: 1, expected: "1s"},
		{name: "one minute", input: 60, expected: "1m 0s"},
		{name: "one hour", input: 3600, expected: "1h 0m 0s"},
		{name: "one day", input: 86400, expected: "1d 0h 0m 0s"},
		{name: "one day and one hour", input: 86400 + 3600, expected: "1d 1h 0m 0s"},
		{name: "negative duration", input: -(86400*2 + 3600*3 + 60*4 + 5), expected: "-2d 3h 4m 5s"},
		{name: "Using a float", input: 899.99, expected: "14m 59s"},
	}
	
	for _, tt := range tc {
		t.Run(tt.name, func(t *testing.T) {
			result, err := HumanizeDuration(tt.input)
			require.NoError(t, err)
			require.Equal(t, tt.expected, result)
		},
	}

```

The same comment applies for all of the tests in the file.

File: helpers/templates/convert_to_float.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+package templates
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+)
+
+func convertToFloat(i interface{}) (float64, error) {
```
Comment (gotjosh) [inline helpers/templates/convert_to_float.go:22]: I've verified this and is indeed a copy, thanks.

File: helpers/templates/humanize_duration.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
+// Copyright 2015 The Prometheus Authors
```
Comment (gotjosh) [inline helpers/templates/humanize_duration.go:1]:
```suggestion
// Copyright 2024 The Prometheus Authors
```

File: helpers/templates/humamize_duration_test.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
+// Copyright 2015 The Prometheus Authors
```
Comment (gotjosh) [inline helpers/templates/humamize_duration_test.go:1]:
```suggestion
// Copyright 2024 The Prometheus Authors
```

**Review by gotjosh — CHANGES_REQUESTED** [review]
Looks great @freak12techno - apologies for the wait 🙏 , I've dropped some comments but all of them a very minor. If you could please address them I'd be thankful if not, I'll do it myself and merge this - just let me know.

**Comment by freak12techno** [comment]
@gotjosh done, can you review?

**Review by gotjosh — APPROVED** [review]
LGTM

Lovely stuff, appreciate your quick turn around .

**Comment by gotjosh** [comment]
@freak12techno thank you very much for your contribution 🙏

**Comment by gotjosh** [comment]
@ArthurSens can you help me bump a new patch version? Is it the same process as other repositories? TIA.

**Comment by gotjosh** [comment]
I think I've figured it out - you should be good to go @freak12techno.

**Comment by ArthurSens** [comment]
Yeah, there's no release process here as far as I know. Just go to the release pages and do things manually :)

**Pull request merged**